lib/bap/bap_synthetic_symbolizer.ml

```diff
@@ -0,0 +1,142 @@
+open Core_kernel.Std
+open Bap_types.Std
+open Bap_image_std
+open Bap_disasm_std
+open Bap_sema.Std
+open Monads.Std
+
+include Bap_self.Create()
+
+module Fact = Ogre.Make(Monad.Ident)
+
+module Ext = struct
+  open Image.Scheme
+  open Fact.Syntax
+
+  let addr_width =
+    Fact.require arch >>= fun a ->
+    match Arch.of_string a with
+    | Some a -> Fact.return (Arch.addr_size a |> Size.in_bits)
+    | None -> Fact.failf "unknown/unsupported architecture" ()
+
+  let calls insns =
+    let insns =
+      Seq.filter insns ~f:(fun (mem, insn) ->
+          Bap_disasm_insn.(may affect_control_flow insn)) in
+    addr_width >>= fun width ->
+    Fact.collect
+      Ogre.Query.(select (from external_reference)) >>= fun syms ->
+    let tab = Seq.fold ~init:Bap_table.empty ~f:(fun tab (mem,_) ->
```

Comment:
two points here: first, do not name variables by their type, e.g., a table, a list, etc, unless you're writing a generic algorithm for that particular container. Name variables according to what they represent, not according to how the represent it. The second issue, is that it is much more readable when you put the value that you're folding near the `fold` function, not like 10 lines later. Basically, it should be easily deductible from code, that this code iterates over `insns`.